File: core/src/main/kotlin/com/example/toyracers/track/Track.kt
package com.example.toyracers.track

/** Immutable data needed to simulate and draw one race track. */
data class Track(
    val id: String,
    val name: String,
    val worldBounds: TrackRectangle,
    val cameraBounds: TrackRectangle,
    val outerBoundary: TrackRectangle,
    val innerObstacles: List<TrackRectangle>,
    val collisionShapes: List<TrackCollisionShape> = emptyList(),
    val backgroundSurface: SurfaceType,
    val surfaceRegions: List<SurfaceRegion>,
    val roadRegion: StadiumRing? = null,
    val roadOuter: TrackPolygon? = null,
    val roadInner: TrackPolygon? = null,
    val startLine: StartLine,
    val checkpoints: List<Checkpoint>,
    val startGrid: List<StartGridPosition>,
    val racingLine: List<TrackPoint>,
    val racingLineWaypointRadius: Float = 3f,
) {
    init {
        require(id.isNotBlank()) { "Track id must not be blank" }
        require(name.isNotBlank()) { "Track name must not be blank" }
        require(worldBounds.contains(cameraBounds)) {
            "Camera bounds must be inside world bounds"
        }
        require(worldBounds.contains(outerBoundary)) {
            "Outer boundary must be inside world bounds"
        }
        require(innerObstacles.all(outerBoundary::contains)) {
            "Inner obstacles must be inside the outer boundary"
        }
        require(surfaceRegions.all { worldBounds.contains(it.bounds) }) {
            "Surface regions must be inside world bounds"
        }
        require(innerObstacles.all(worldBounds::contains)) {
            "Collision obstacles must be inside world bounds"
        }
        require(outerBoundary.contains(startLine.bounds)) {
            "Start line must be inside the outer boundary"
        }
        require(checkpoints.isNotEmpty()) { "Track must have checkpoints" }
        require(checkpoints.map(Checkpoint::order) == checkpoints.indices.toList()) {
            "Checkpoints must be ordered and use contiguous indices"
        }
        require(
            checkpoints.all {
                worldBounds.contains(it.gate.start) && worldBounds.contains(it.gate.end)
            },
        ) {
            "Checkpoint gates must be inside world bounds"
        }
        require(startGrid.isNotEmpty()) { "Track must have start positions" }
        require(startGrid.all { worldBounds.contains(it.position) }) {
            "Start positions must be inside world bounds"
        }
        require(racingLine.size >= MIN_RACING_LINE_POINTS) {
            "Racing line must contain at least $MIN_RACING_LINE_POINTS points"
        }
        require(racingLine.all(worldBounds::contains)) {
            "Racing line must be inside world bounds"
        }
        require(racingLineWaypointRadius > 0f) {
            "Racing line waypoint radius must be positive"
        }
        require((roadOuter == null) == (roadInner == null)) {
            "Tiled road contours must be provided together"
        }
    }

    fun surfaceAt(point: TrackPoint): SurfaceType {
        return surfaceAt(point.x, point.y)
    }

    fun surfaceAt(
        x: Float,
        y: Float,
    ): SurfaceType {
        if (innerObstacles.any { it.contains(x, y) }) {
            return backgroundSurface
        }
        val insideTiledRoad =
            roadOuter?.contains(x, y) == true && roadInner?.contains(x, y) == false
        if (insideTiledRoad || roadRegion?.contains(x, y) == true) {
            return SurfaceType.ASPHALT
        }
        return surfaceRegions.lastOrNull { it.bounds.contains(x, y) }?.surface
            ?: backgroundSurface
    }

    private companion object {
        const val MIN_RACING_LINE_POINTS = 3
    }
}

data class SurfaceRegion(
    val bounds: TrackRectangle,
    val surface: SurfaceType,
)

data class StartLine(
    val bounds: TrackRectangle,
    val forwardX: Float,
    val forwardY: Float,
) {
    init {
        require(forwardX != 0f || forwardY != 0f) {
            "Start line forward direction must not be zero"
        }
    }
}

File: dart/lib/simulation/track/track.dart
import '../math/float32.dart';
import '../surface/surface_type.dart';
import 'checkpoint.dart';
import 'start_grid_position.dart';
import 'track_geometry.dart';
import 'track_point.dart';

/// Immutable simulation data for one complete race track.
final class Track {
  Track({
    required String id,
    required String name,
    required this.worldBounds,
    required this.cameraBounds,
    required this.outerBoundary,
    Iterable<TrackRectangle> innerObstacles = const <TrackRectangle>[],
    Iterable<TrackCollisionShape> collisionShapes =
        const <TrackCollisionShape>[],
    required this.backgroundSurface,
    Iterable<SurfaceRegion> surfaceRegions = const <SurfaceRegion>[],
    this.roadOuter,
    this.roadInner,
    required this.startLine,
    required Iterable<Checkpoint> checkpoints,
    required Iterable<StartGridPosition> startGrid,
    required Iterable<TrackPoint> racingLine,
    double racingLineWaypointRadius = 3,
  }) : id = _requireText(id, 'id'),
       name = _requireText(name, 'name'),
       innerObstacles = List<TrackRectangle>.unmodifiable(innerObstacles),
       collisionShapes = List<TrackCollisionShape>.unmodifiable(
         collisionShapes,
       ),
       surfaceRegions = List<SurfaceRegion>.unmodifiable(surfaceRegions),
       checkpoints = List<Checkpoint>.unmodifiable(checkpoints),
       startGrid = List<StartGridPosition>.unmodifiable(startGrid),
       racingLine = List<TrackPoint>.unmodifiable(racingLine),
       racingLineWaypointRadius = Float32.narrow(racingLineWaypointRadius) {
    if (!worldBounds.containsRectangle(cameraBounds)) {
      throw ArgumentError('Camera bounds must be inside world bounds.');
    }
    if (!worldBounds.containsRectangle(outerBoundary)) {
      throw ArgumentError('Outer boundary must be inside world bounds.');
    }
    if (!innerObstacles.every(outerBoundary.containsRectangle)) {
      throw ArgumentError('Inner obstacles must be inside the outer boundary.');
    }
    if (!surfaceRegions.every(
      (region) => worldBounds.containsRectangle(region.bounds),
    )) {
      throw ArgumentError('Surface regions must be inside world bounds.');
    }
    if (!innerObstacles.every(worldBounds.containsRectangle)) {
      throw ArgumentError('Collision obstacles must be inside world bounds.');
    }
    if (!outerBoundary.containsRectangle(startLine.bounds)) {
      throw ArgumentError('Start line must be inside the outer boundary.');
    }
    if (checkpoints.isEmpty) {
      throw ArgumentError('Track must have checkpoints.');
    }
    if (!_hasOrderedCheckpoints(this.checkpoints)) {
      throw ArgumentError('Checkpoints must use contiguous ordered indices.');
    }
    if (!this.checkpoints.every(
      (checkpoint) =>
          worldBounds.containsPoint(checkpoint.gate.start) &&
          worldBounds.containsPoint(checkpoint.gate.end),
    )) {
      throw ArgumentError('Checkpoint gates must be inside world bounds.');
    }
    if (this.startGrid.isEmpty) {
      throw ArgumentError('Track must have start positions.');
    }
    if (!this.startGrid.every(
      (position) => worldBounds.containsPoint(position.position),
    )) {
      throw ArgumentError('Start positions must be inside world bounds.');
    }
    if (this.racingLine.length < _minimumRacingLinePoints) {
      throw ArgumentError.value(
        this.racingLine.length,
        'racingLine',
        'must contain at least $_minimumRacingLinePoints points',
      );
    }
    if (!this.racingLine.every(worldBounds.containsPoint)) {
      throw ArgumentError('Racing line must be inside world bounds.');
    }
    if (this.racingLineWaypointRadius <= 0) {
      throw ArgumentError.value(
        racingLineWaypointRadius,
        'racingLineWaypointRadius',
        'must be greater than zero',
      );
    }
    if ((roadOuter == null) != (roadInner == null)) {
      throw ArgumentError('Tiled road contours must be provided together.');
    }
  }

  static const int _minimumRacingLinePoints = 3;

  final String id;
  final String name;
  final TrackRectangle worldBounds;
  final TrackRectangle cameraBounds;
  final TrackRectangle outerBoundary;
  final List<TrackRectangle> innerObstacles;
  final List<TrackCollisionShape> collisionShapes;
  final SurfaceType backgroundSurface;
  final List<SurfaceRegion> surfaceRegions;
  final TrackPolygon? roadOuter;
  final TrackPolygon? roadInner;
  final StartLine startLine;
  final List<Checkpoint> checkpoints;
  final List<StartGridPosition> startGrid;
  final List<TrackPoint> racingLine;
  final double racingLineWaypointRadius;

  /// Returns the surface selected by the reference track rules at [point].
  SurfaceType surfaceAt(TrackPoint point) =>
      surfaceAtCoordinates(point.x, point.y);

  /// Returns the surface selected by the reference track rules at [x], [y].
  SurfaceType surfaceAtCoordinates(double x, double y) {
    final pointX = Float32.narrow(x);
    final pointY = Float32.narrow(y);
    if (innerObstacles.any((bounds) => bounds.contains(pointX, pointY))) {
      return backgroundSurface;
    }
    final insideTiledRoad =
        roadOuter?.contains(pointX, pointY) == true &&
        roadInner?.contains(pointX, pointY) == false;
    if (insideTiledRoad) {
      return SurfaceType.asphalt;
    }
    for (var index = surfaceRegions.length - 1; index >= 0; index--) {
      final region = surfaceRegions[index];
      if (region.bounds.contains(pointX, pointY)) {
        return region.surface;
      }
    }
    return backgroundSurface;
  }

  static bool _hasOrderedCheckpoints(List<Checkpoint> checkpoints) {
    for (var index = 0; index < checkpoints.length; index++) {
      if (checkpoints[index].order != index) {
        return false;
      }
    }
    return true;
  }

  static String _requireText(String value, String name) {
    if (value.trim().isEmpty) {
      throw ArgumentError.value(value, name, 'must not be blank');
    }
    return value;
  }
}

/// An ordered rectangular override of a track's background surface.
final class SurfaceRegion {
  const SurfaceRegion({required this.bounds, required this.surface});

  final TrackRectangle bounds;
  final SurfaceType surface;
}

/// The gate that completes a lap when crossed in its forward direction.
final class StartLine {
  StartLine({
    required this.bounds,
    required double forwardX,
    required double forwardY,
  }) : forwardX = Float32.narrow(forwardX),
       forwardY = Float32.narrow(forwardY) {
    if (this.forwardX == 0 && this.forwardY == 0) {
      throw ArgumentError('Start line forward direction must not be zero.');
    }
  }

  final TrackRectangle bounds;
  final double forwardX;
  final double forwardY;
}
